libyul/optimiser/SSAReverser.cpp

```diff
@@ -29,42 +29,74 @@ void SSAReverser::operator()(Block& _block)
 		_block.statements,
 		[&](Statement& _stmt1, Statement& _stmt2) -> boost::optional<vector<Statement>>
 		{
+			auto* varDecl = boost::get<VariableDeclaration>(&_stmt1);
+
+			if (!varDecl || varDecl->variables.size() != 1 || !varDecl->value)
+				return {};
+
 			// Replaces
 			//   let a_1 := E
 			//   a := a_1
 			// with
 			//   a := E
 			//   let a_1 := a
-
-			auto* varDecl = boost::get<VariableDeclaration>(&_stmt1);
-			auto* assignment = boost::get<Assignment>(&_stmt2);
-
-			if (!varDecl || !assignment)
-				return {};
-
-			auto* identifier = boost::get<Identifier>(assignment->value.get());
-
-			if (
-				varDecl->variables.size() == 1 &&
-				varDecl->value &&
-				assignment->variableNames.size() == 1 &&
-				identifier &&
-				identifier->name == varDecl->variables.front().name
-			)
+			if (auto* assignment = boost::get<Assignment>(&_stmt2))
+			{
+				auto* identifier = boost::get<Identifier>(assignment->value.get());
+				if (
+					assignment->variableNames.size() == 1 &&
+					identifier &&
+					identifier->name == varDecl->variables.front().name
+					)
+				{
+					vector<Statement> result;
+					result.emplace_back(Assignment{
+						std::move(assignment->location),
+						assignment->variableNames,
+						std::move(varDecl->value)
+					});
+					result.emplace_back(VariableDeclaration{
+						std::move(varDecl->location),
+						std::move(varDecl->variables),
+						std::make_unique<Expression>(std::move(assignment->variableNames.front()))
+					});
+					return { std::move(result) };
+				}
+			}
+			// Replaces
+			//   let a_1 := E
+			//   let a := a_1
+			// with
+			//   let a := E
+			//   let a_1 := a
+			else if (auto* varDecl2 = boost::get<VariableDeclaration>(&_stmt2))
 			{
-				vector<Statement> result;
-				result.emplace_back(Assignment{
-					std::move(assignment->location),
-					assignment->variableNames,
-					std::move(varDecl->value)
-				});
-				result.emplace_back(VariableDeclaration{
-							std::move(varDecl->location),
-							std::move(varDecl->variables),
-							std::make_unique<Expression>(std::move(assignment->variableNames.front()))
-				});
-				return { std::move(result) };
+				auto* identifier = boost::get<Identifier>(varDecl2->value.get());
+				if (
+					varDecl2->variables.size() == 1 &&
+					identifier &&
+					identifier->name == varDecl->variables.front().name
```

Comment:
To make the transformation "stable", you could add the condition that `varDecl2->variables.front().name.str().size() > varDecl1->variables.front().name.str().size()`.

Reply:
Is it unstable otherwise? I'll need to think about that :-). It might be that it can never switch back anyways due to variable visibility...

Reply:
Yes, it's indeed unstable... hm... I need to look at this more carefully once more - I'm actually not sure anymore why the second back transformations seems to be beneficial at all, maybe something is wrong there...

Reply:
Ah ok... the foremost criterion should probably be to move the one that is assigned to more often later up... and if they are assigned to equally often, we can skip the transformation altogether...

Reply:
Exactly, if it indeed is SSA, then one of them is never assigned. You can check using the SSAValueTracker.